### PR TITLE
Rewrite guessing vanilla version

### PR DESCRIPTION
**Pull request by selaux**
This rewrites guessing vanilla version to some degree. Motivation for the change is mostly that I think it never worked as intended on my machine (because of not matching paths case insensitively). I found the logic hard to follow as well. 

### Major Changes
- When building resource packs, resources are now read in parallel using rayon
- Iterators and HashMap are used instead of for loops
- Case insensitivity has been introduced
- When comparing resource packs, we first check for language specific resources (e.g. `german` subdirectory), this is just faster (at least on my machine) and less error prone. It will return a game version if ONLY files for this version are found
- We compare resource packs second, in parallel. We calculate percentages for each resource pack, e.g. how many files are missing in data dir that are in the resource pack. We then use the minimal mismatching game version

### Todo

- [x] Use `Nfc` for resource path

### CODE REVIEW

File: rust/src/guess.rs [hunk 1]
```diff
@@ -105,64 +154,90 @@ impl Guess {
...
-                best_num = num_polish;
-                best_version = Some(VanillaVersion::POLISH);
-            }
+
+        let version = match (num_dutch, num_german, num_italian, num_polish, num_russian) {
+            (n, 0, 0, 0, 0) if n > 0 => Some(VanillaVersion::DUTCH),
+            (0, n, 0, 0, 0) if n > 0 => Some(VanillaVersion::GERMAN),
+            (0, 0, n, 0, 0) if n > 0 => Some(VanillaVersion::ITALIAN),
+            (0, 0, 0, n, 0) if n > 0 => Some(VanillaVersion::POLISH),
+            (0, 0, 0, 0, n) if n > 0 => Some(VanillaVersion::RUSSIAN),
```
Comment (lynxlynxlynx) [inline rust/src/guess.rs]: RUSSIAN_GOLD is not relevant for this? I noticed it's not in the old code either, but better ask.

Comment (selaux) [reply rust/src/guess.rs]: Tbh, I just reused the code that was there before. I'm not even sure what the `RUSSIAN_GOLD` version exactly refers to. It seems to work with the russian version from steam (I didn't do exhaustive testing wether it crashes somewhere in the game though)

Comment (flaviojs) [reply rust/src/guess.rs]: That version guessing code is based on the paths returned by this function:
https://github.com/ja2-stracciatella/ja2-stracciatella/blob/c3e9284a195ef38d774ecb4b3c0cf601576cb996/src/game/GameRes.cc#L156

File: rust/src/guess.rs [hunk 2]
```diff
@@ -105,64 +161,95 @@ impl Guess {
...
+    }
+
+    /// Guess version by building resource pack and comparing it to other resource
+    /// packs in externalized directory
+    fn guess_by_resource_matching(&self, datadir: &Path) -> GuessResult<Option<VanillaVersion>> {
+        let results = self
+            .get_pack_paths()?
+            .par_iter()
+            .map(|path| self.compare_pack(datadir, path))
+            .collect::<Result<Vec<_>, _>>()?;
```
Comment (flaviojs) [inline rust/src/guess.rs]: This `collect::<Result<Vec<_>, _>>()?` was confusing me.

I would add a comment saying that it stops early and returns an error (non-deterministic) if any of the parallel calls return an error, or just let everything finish and return a combined error message in case of multiple errors (see [`partition_map`](https://docs.rs/rayon/1.1.0/rayon/iter/trait.ParallelIterator.html#method.partition_map)).

File: rust/src/guess.rs [hunk 3]
```diff
@@ -105,64 +161,95 @@ impl Guess {
...
+            (0, 0, 0, 0, n) if n > 0 => Some(VanillaVersion::RUSSIAN),
+            _ => None,
+        };
+
+        match version {
+            Some(version) => info!(
+                "Detected exclusively language specific files for language {:?}",
+                version
+            ),
+            None => info!("Detected no language specific files or multiple different languages"),
```
Comment (flaviojs) [inline rust/src/guess.rs]: I would list of languages it detected.

Comment (selaux) [reply rust/src/guess.rs]: Added a log line for that.

File: rust/src/guess.rs [hunk 4]
```diff
@@ -36,63 +130,25 @@ pub struct Guess {
...
-        info!("Inspecting resource paths");
+        info!("Inspecting resource paths for language specific files");
         let mut num_dutch = 0;
         let mut num_german = 0;
         let mut num_italian = 0;
         let mut num_polish = 0;
         let mut num_russian = 0;
         for resource in resources {
-            let p = resource.path.to_lowercase().replace("\\", "/");
+            let p = Nfc::caseless(&resource.path);
```
[the remark this answers is not in the document]
Comment (selaux) [reply rust/src/guess.rs]: I added tests for case insensitivity.

File: rust/src/guess.rs (hunk 2 above)
Comment (selaux) [reply rust/src/guess.rs]: I added a comment to that. We don't have a way to return multiple errors from that function yet so `partition_map` seems like it would require quite some overhead (collecting all errors and somehow joining them into a single string).

File: rust/src/guess.rs [hunk 5]
```diff
@@ -1,6 +1,9 @@
 //! This module contains code to guess Vanillaversion with resource packs.
 
 use log::{error, info};
+use rayon::iter::IntoParallelRefIterator;
+use rayon::iter::ParallelIterator;
```
[the remark this answers is not in the document]
Comment (selaux) [reply rust/src/guess.rs]: Done.

**Comment by selaux** [comment]
I addressed the comments...
